zetteldeft.el

```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
-          (deft-dir (file-name-as-directory deft-directory))
-          new-filename old-name new-name)
-      (when old-filename
-        (setq old-name (deft-base-filename old-filename))
-        (setq new-name (read-string
-                        (concat "Rename " old-name " to (without extension): ")
-                        old-name))
-        (setq new-filename
-              (concat deft-dir new-name "." deft-default-extension))
-        (rename-file old-filename new-filename)
-        (deft-update-visiting-buffers old-filename new-filename)
-        (zetteldeft-update-title-in-file)
-        (deft-refresh))))
-
-(defun zetteldeft-update-title-in-file ()
+  (let ((old-filename (buffer-file-name))
+	prompt-pars new-name new-filename)
+    (when old-filename
+      (setq prompt-pars (funcall cusprompt old-filename))
+      (setq new-name (read-string (car prompt-pars) (cdr prompt-pars)))
+      (setq new-filename (funcall fn-gen old-filename new-name))
+      (rename-file old-filename new-filename)
+      (deft-update-visiting-buffers old-filename new-filename)
+      (if update-title (zetteldeft-update-title-in-file new-name) nil)
+      (deft-refresh))))
+
+(defun zetteldeft-update-title-in-file (title)
   "Update the title of the current file, if present.
-Does so by looking for `zetteldeft-title-prefix'."
+Does so by looking for `zetteldeft-title-prefix'.
+Replaces current title with TITLE."
   (save-excursion
     (let ((zetteldeft-title-suffix ""))
       (goto-char (point-min))
       (when (re-search-forward (regexp-quote zetteldeft-title-prefix) nil t)
         (delete-region (line-beginning-position) (line-end-position))
-        (zetteldeft--insert-title)))))
+        (zetteldeft--insert-title title)))))
+
+(defun zetteldeft-file-rename ()
+  "Change current file's title, and use the new title to rename the file."
+  (interactive)
+  (zetteldeft--rename
+   (lambda (old-filename) (let (old-title prompt-text)
+			    (setq old-title (zetteldeft--lift-file-title old-filename))
+			    (setq prompt-text (concat "Change " old-title " to: "))
+			    `((,@prompt-text) (,@old-title))))
+   (lambda (old-filename new-name) (let (old-id old-extension new-filename)
+				     (setq old-id (zetteldeft--lift-id old-filename))
+				     (setq old-extension (zetteldeft--identify-extension old-filename))
+				     (setq new-filename (concat (file-name-sans-extension (deft-absolute-filename (concat old-id zetteldeft-id-filename-separator new-name))) "." old-extension))
+				     new-filename))
+   t))
+
+(defun zetteldeft--identify-extension (filename)
+  "Identify the extension of the provided FILENAME.
+Handles multiple file extensions, if extension exists in deft-extensions."
+  (let (extension)
+      (setq extension (-reduce-from (lambda (a x) (if (string-match (concat x "$") filename) x a)) nil (--sort (< (length it) (length other)) deft-extensions)))
+      (if extension extension (file-name-extension filename))))
```

Comment:
Is this necessary? This seems to redefine the concept of a file extension, which is traditionally only the final dotted segment. For instance, the file extension for `archive.tar.gz` is `gz`; the string ".tar" is part of the file name.

At the very least, this feature seems unrelated to this PR and better as a separate issue/commit.

Reply:
My driver to add this feature was to support literate Agda with the file extension `.lagda.org`. The default behaviour is still to only use the final dotted segment, it just allows for a "custom" file extension to be supported if it's added to `deft-extensions`.

Reply:
Wouldn't it make more sense to use a supported file extension like `.lagda-org` and associate your desired major mode via `auto-mode-alist` in your `init.el` file? If you try to redefine what a file extension is, it's not just zetteldeft you'd have to change; it's emacs own behavior, and any other package you might install.

Reply:
I wish it was that easy:
```
test.lagda-org: Unsupported extension.
Supported extensions are: .agda, .lagda, .lagda.rst, .lagda.tex,
                          .lagda.md, .lagda.org
```
The major mode integrates with the compiler, and its literate mode uses those extensions.

Reply:
Why not use `.lagda` then? It's supported by your compiler and is, without ambiguity, a file extension.

Reply:
Sorry, I didn't put enough information in my last post. The `.lagda` extension is for TeX, the only way the compiler appears to support org mode is with `.lagda.org`. I found the source [here](https://github.com/agda/agda/blob/32664c045fffd6708c2d46204a70fcd48f063fe3/src/full/Agda/Syntax/Parser/Literate.hs#L92-L100)

I've had another idea, will post in the main thread below.

Reply:
Ah, got it. That source clarifies things.

Out of curiosity, have you tried Agda with `org-babel`?